AzureMonitor: Add support for not equals and startsWith operators when creating Azure Metrics dimension filters

aangelisc opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

Adds support for the not equals and startsWith operators in the Azure Monitor data source when selecting dimensions for a metric (where supported).

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #47427 

**Special notes for your reviewer**:

aangelisc committed:
Allow dimension operator selection

- Add dimension operators and function to update the operator in the query
- Add logic to ensure the same dimension cannot be selected multiple times (Azure restriction)
- Add selection component
- Remove default operator setting to 'eq'

aangelisc committed:
Update backend logic to default operation and filter to eq '*'

- This must be done as the ne and sw operators do not work with the wildcard filter

aangelisc committed:
Add tests on dimension operators

aangelisc committed:
Remove dimension options filtering (azure does allow multiple select)

aangelisc committed:
Remove duplicate test

aangelisc committed:
Readd dimension filtering as Azure only seems to support this in portal

iwysiu commented on pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go:
```diff
@@ -120,11 +120,16 @@ func (e *AzureMonitorDatasource) buildQueries(queries []backend.DataQuery, dsInf
+					dimSB.WriteString(fmt.Sprintf("%s eq '*'", filter.Dimension))
+				} else {
+					dimSB.WriteString(filter.String())
+					if i != len(azJSONModel.DimensionFilters)-1 {
```
should the code that adds `" and "` be outside this else?

aangelisc replied:
Great catch, missed this 😊 Explains why the below test was passing when it shouldn't of!

iwysiu commented on pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource_test.go:
```diff
@@ -127,6 +127,39 @@ func TestAzureMonitorBuildQueries(t *testing.T) {
+			},
+			queryInterval:           duration,
+			expectedInterval:        "PT1M",
+			azureMonitorQueryTarget: "%24filter=blob+eq+%27%2A%27tier+eq+%27%2A%27&aggregation=Average&api-version=2018-01-01&interval=PT1M&metricnames=Percentage+CPU&metricnamespace=Microsoft.Compute-virtualMachines&timespan=2018-03-15T13%3A00%3A00Z%2F2018-03-15T13%3A34%3A00Z&top=30",
```
See azuremonitor-datasource.go, but should this be:
`%24filter=blob+eq+%27%2A%27+and+tier+eq+%27%2A%27&a...` ?

aangelisc replied:
Fixed as per above

iwysiu commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -15,6 +16,21 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+    let t = dimensionOptions;
+    let dimensionFilters = query.azureMonitor?.dimensionFilters;
+    if (dimensionFilters !== undefined && dimensionFilters.length > 0) {
+      t = dimensionOptions.filter((val) => !dimensionFilters?.find((dimension) => dimension.dimension === val.value));
```
Maybe add a comment? This bit of logic is a little complex

sunker replied:
A few nits: 
An undefined dimensionFilters and an empty array both evaluates to false, so you should be able to replace the if condition to `if (query.azureMonitor?.dimensionFilters?.length) {`. 

Think you can also remove the not-null assertion operator from the filter condition. Using the `dimensionFilters?.find` is fine, but using `dimensionFilters?.some` would be a bit more expressive in this case. 

 Maybe a unit test can be added for this?

iwysiu commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -46,11 +62,17 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
               width={38}
             />
-            <InlineLabel aria-label="equals">==</InlineLabel>
+            <Select
```
optional, but it might be nice to set `==` as the default value

aangelisc replied:
This makes sense given that in the backend we default to `==` as well.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -46,11 +62,17 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+              value={filter.operator}
+              options={dimensionOperators}
+              onChange={(v) => onFieldChange(index, 'operator', v.value ?? '')}
+            />
```
If you want to, you can take a look at how operators are rendered in the CloudMonitoring filter component. There the input control is replaced by a button. I think it looks pretty good in the scenarios where predefined non text values such as operators are listed. I think it would be nice if we could do the same here.

![Screenshot from 2022-04-22 09-23-04](https://user-images.githubusercontent.com/2388950/164627712-25b89e76-6791-4f3e-91cd-56fb87d63ce6.png)

https://github.com/grafana/grafana/blob/47427-additional-dimension-operations/public/app/plugins/datasource/cloud-monitoring/components/LabelFilter.tsx#L126

aangelisc replied:
This makes sense for consistency, will take a look at implementing this 😊

aangelisc replied:
On the question of creating a custom operator value if Azure adds a new operator - I don't believe this would work as the display name of the operator doesn't directly match up with the actual value (e.g. `==` in the frontend is translated to `eq`). We could alternatively use the actual API representation however this may not be as appealing. It wouldn't require a change in the backend as we set the appropriate API value in the frontend here: 
![image](https://user-images.githubusercontent.com/15019026/164699132-240a3bb5-cff0-4db9-8109-f8524876a582.png)

sunker replied:
Maybe for optional values, it could be fine that the label name and value are the same? `{ label: 'ew', 'value: 'ew }` Thoughts?

aangelisc replied:
I think that'd be a happy medium as, if the user is specifying the operator, the correct value should be passed to the backend.

sunker reviewed (commented):
This looks really promising @aangelisc, great work! 

A question: React select has the option to `allowCustomValue` which gives the user the freedom to add whatever value the want. Could that be an option here so that we provide a workaround in case Azure decides to add support for new operators? Or would that still require manual changes in the backend?

aangelisc commented on pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go:
```diff
@@ -120,11 +120,16 @@ func (e *AzureMonitorDatasource) buildQueries(queries []backend.DataQuery, dsInf
+					}
 				}
 			}
+
```
👍

aangelisc committed:
Correct placement of 'and' when building query

aangelisc committed:
Add comment and simplify filtering logic

aangelisc committed:
Allow multiSelect for eq and ne operators

- Pass PanelData to DimensionFields component
- Add logic to retrieve labels from PanelData
- Add MultiSelect component for relevant operators
- Update frontend types to allow filter to be an array of strings
- Update backend types to allow filter to be an array of strings
- Update filter string building

aangelisc committed:
Improve setting of labels

aangelisc committed:
Update go tests

aangelisc committed:
Update frontend tests

- Add panelData mock (to be expanded later)
- Update null check in DimensionFields

aangelisc committed:
Allow custom value and set default

dprokop reviewed (commented):
Think this PR would use a rebase: 
<img width="226" alt="image" src="https://user-images.githubusercontent.com/2376619/165072262-8a451c1a-f487-4861-aee2-63d663a900ff.png">

aangelisc committed:
Merge remote-tracking branch 'origin/main' into 47427-additional-dimension-operations

# Conflicts:
#	public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx
#	public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricsQueryEditor.test.tsx

aangelisc committed:
Add frontend test and fix lint issues

fridgepoet commented on pkg/tsdb/azuremonitor/types/types.go:
```diff
@@ -139,15 +140,27 @@ type AzureMonitorJSONQuery struct {
-	Filter    string `json:"filter"`
+	Dimension string   `json:"dimension"`
+	Operator  string   `json:"operator"`
+	Filter    []string `json:"filter"`
```
From our conversation: suggested moving this change to a separate PR

sunker replied:
Yeah I'm not sure I follow. Why is this needed? Unless migrated properly in frontend and backend, it will break any existing queries passing `Filter` as a string. Would be good with a new issue that provides context on why this change is necessary.

fridgepoet commented on pkg/tsdb/azuremonitor/types/types.go:
```diff
@@ -139,15 +140,27 @@ type AzureMonitorJSONQuery struct {
 	filter := "*"
-	if a.Filter != "" {
-		filter = a.Filter
+	if len(a.Filter) >= 1 {
+		if len(a.Filter) == 1 {
+			filter = a.Filter[0]
+		} else {
+			var filterStrings []string
+			for _, filter := range a.Filter {
+				filterStrings = append(filterStrings, fmt.Sprintf("%v %v '%v'", a.Dimension, a.Operator, filter))
+			}
+			if a.Operator == "eq" {
+				return strings.Join(filterStrings, " or ")
+			} else {
+				return strings.Join(filterStrings, " and ")
+			}
+		}
```
from our conversation: suggest moving this to a different function more focused on filter construction, will be able to unit test it closer to its declaration

aangelisc replied:
Will move this to a separate PR

aangelisc committed:
Improved handling of options for sw operator

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -105,11 +105,11 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
       .filter((f) => f.dimension && f.dimension !== 'None')
       .map((f) => {
-        const filter = templateSrv.replace(f.filter ?? '', scopedVars);
+        const filter = f.filter?.map((filter) => templateSrv.replace(filter ?? '', scopedVars)).filter((item) => item!);
```
Can this last filter be moved to the already existing filter on line 106 so that we don't iterate over all items another time?

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -105,11 +105,11 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
           dimension: templateSrv.replace(f.dimension, scopedVars),
           operator: f.operator || 'eq',
-          filter: filter || '*', // send * when empty
+          filter: filter, // send * when empty
```
comment not useful anymore so can be removed?

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -11,12 +12,60 @@ interface DimensionFieldsProps extends AzureQueryEditorFieldProps {
+        }
+      }
+    }
+    setDimensionLabels((prevLabels) => ({ ...labelsObj, ...prevLabels }));
```
If I understand correctly, it extracts the labels from all the returned series and put them in a multi select filter? I'm not sure this is correct. data.series returns labels for all queries in a panel - not only for the current query row. To be sure you use labels for the current query, you need to check the frame's refId against the row that is currently being rendered in the editor. 

I think this hook is a bit complicated. Would be nice if you could add some comments on what it does. Also, make sure all args in the arrow functions are properly named. For example `fields.map((item) => item.labels)` could be `fields.map((field) => field.labels)` to make it easier to read the code. 

You can also consider 1) moving this to a custom hook or 2) moving the multi-select filter to a separate component that would also contain this hook.

aangelisc replied:
Great catch on needing to check the refId, will update this.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -33,12 +82,22 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
-      onFieldChange(index, 'filter', ev.target.value);
+  const onFilterInputChange = (index: number, v: SelectableValue<string>) => {
+    if (v) {
+      onFieldChange(index, 'filter', [v.value!]);
```
We should avoid using the non-null assertion operator. If the value is nullable, we should check that this is not the case.

aangelisc replied:
Corrected, thanks 😊

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -33,12 +82,22 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+    onFieldChange(
+      index,
+      'filter',
+      v.map((item) => item.value!)
```
same here

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/types/query.ts:
```diff
@@ -87,5 +87,5 @@ export interface AzureResourceGraphQuery {
   dimension: string;
   operator: string;
-  filter?: string;
+  filter?: string[];
```
This is problematic. If we don't migrate existing queries, they will still send strings to the backend. If we for some reason don't want to migrate queries in the frontend but rely on backend to handle both, we should at least change typings here to `string | string[]`.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -48,12 +107,44 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+              allowCustomValue
+            />
+            {filter.operator === 'eq' || filter.operator === 'ne' ? (
+              <MultiSelect
```
Can you add unit tests for the conditional rendering?

aangelisc replied:
Will be moved to a separate PR

sunker reviewed (commented):
Added a few more comments. Agree with @fridgepoet - the latest changes to the query model should go in a separate PR.

aangelisc committed:
Remove changes related to multiselect

aangelisc committed:
Add check on refId to ensure dimension labels are correct for query

aangelisc committed:
Extract custom hook for setting dimension labels

aangelisc committed:
Add documentation around Azure Monitor metrics dimensions

Commit:
Merge remote-tracking branch 'origin/main' into 47427-additional-dimension-operations

Commit:
Update MetricQueryEditor tests

- Add missing data prop

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=48077) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=48077) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: aangelisc<br/>:x: Andreas Christou<br/><hr/>**Andreas Christou** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=48077) it.</sub>

fridgepoet commented on pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource_test.go:
```diff
@@ -130,6 +130,39 @@ func TestAzureMonitorBuildQueries(t *testing.T) {
+			},
+			queryInterval:           duration,
+			expectedInterval:        "PT1M",
+			azureMonitorQueryTarget: "%24filter=blob+ne+%27test%27&aggregation=Average&api-version=2018-01-01&interval=PT1M&metricnames=Percentage+CPU&metricnamespace=Microsoft.Compute-virtualMachines&timespan=2018-03-15T13%3A00%3A00Z%2F2018-03-15T13%3A34%3A00Z&top=30",
```
I was testing it manually (pulled in latest changes from main) and after I clicked on View in Azure Portal, I didn't see the dimension filters I applied there. Just wondering if these expected `azureMonitorQueryTarget` implied I was supposed to see them passed through there.

aangelisc replied:
Great spot here @fridgepoet, I think this is actually a separate issue in how we build the deep link URL for the query. The URL built above is the URL that is used for data retrieval (the query URL). The URL that is built to link back to the Azure portal does not include filters at the moment, maybe something we can look into implementing.

aangelisc committed:
Correctly set field values

iwysiu reviewed (commented):
nice! It looks a lot nicer.
Is `eq` supposed to be an option? I opened a fresh panel and go this as the operator options:
<img width="262" alt="Screen Shot 2022-04-27 at 5 40 15 PM" src="https://user-images.githubusercontent.com/5421859/165636030-f3de1101-1cf4-4c45-a8b4-3f757b1e3cc5.png">

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -33,12 +89,40 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+    if (v) {
+      onFieldChange(index, 'filter', v.value);
+    } else {
+      onFieldChange(index, 'filter', '');
```
Is this necessary? I think the if else condition can be replaced with `v && onFieldChange(index, 'filter', v.value);`

aangelisc replied:
Have replaced this with ```  const onFilterInputChange = (index: number, v: SelectableValue<string> | null) => {
    onFieldChange(index, 'filter', v?.value ?? '');
  };```

It's necessary as a null value is passed if the filters are cleared so we handle that accordingly.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -48,12 +132,28 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+              allowCustomValue
+              options={getValidFilterOptions(filter.filter, filter.dimension)}
+              onChange={(v) => onFilterInputChange(index, v)}
+              isClearable
```
Should `isClearable` allow you to completely clear the select? It wasn't working for me. Probably doing something wrong, but if not this is an issue in grafana ui so nothing to care about in this pr.,

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -48,12 +132,28 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ query, dimensionOptio
+            <Select
+              menuShouldPortal
+              placeholder="Select value"
+              value={filter.filter ? filter.filter : ''}
```
```suggestion
              value={filter.filter ?? ''}
```

sunker reviewed (commented):
Looks good, great stuff @aangelisc! Wasn't able to find a metric that has dimension labels attached to it so couldn't test filter values properly. Let's discuss this offline.

aangelisc commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.test.tsx:
```diff
@@ -0,0 +1,137 @@
+      ...mockQuery,
+      azureMonitor: { ...mockQuery.azureMonitor, dimensionFilters: [{ dimension: '', operator: 'eq', filter: '*' }] },
+    });
+    render(
```
This should rerender otherwise we'd be testing against the original component.

aangelisc commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.test.tsx:
```diff
@@ -0,0 +1,137 @@
+    await selectOptionInTest(dimensionSelect, 'Test Dimension 1');
+    expect(screen.queryByText('Test Dimension 1')).toBeInTheDocument();
+    expect(screen.queryByText('==')).toBeInTheDocument();
+  });
```
Yep! Good shout, have added this in.

aangelisc committed:
Add additional expect for onQueryChange

aangelisc committed:
Correctly set operators

- Simplify onFilterInputChange

aangelisc committed:
Ensure no duplicate filters appear

aangelisc committed:
Ensure that filters are displayed correctly for saved queries

aangelisc committed:
Update dimension filter test

aangelisc committed:
Include additional test around changing dimension labels

aangelisc committed:
Merge remote-tracking branch 'origin/main' into 47427-additional-dimension-operations

sunker commented:
Might be missing something here but I'm still not able to get any filter value options when using the dashboard you shared with me? 

![azuredimension](https://user-images.githubusercontent.com/2388950/165895602-051b6f77-1ead-415f-98eb-9d60d9912eea.gif)

aangelisc commented:
> Might be missing something here but I'm still not able to get any filter value options when using the dashboard you shared with me?
> 
> ![azuredimension](https://user-images.githubusercontent.com/2388950/165895602-051b6f77-1ead-415f-98eb-9d60d9912eea.gif)

Ah thank you for this! It's because of the new ResourcePicker, will push a commit to rectify this now.

aangelisc committed:
Pass panel data through new metrics query editor

sunker reviewed (approved):
Tested and it works really well! Fantastic work on this @aangelisc! Don't forget to add the `add-to-changelog` label. Also, would it be possible to make the PR title a little more informative? This is the text that will end up in the [changelog](https://github.com/grafana/grafana/blob/main/CHANGELOG.md) eventually.

sunker commented:
@aangelisc the automatic backport job failed, probably due to a merge conflict, so it needs to be done manually. See instructions on how to do it manually in comment above. Let me know if you need help!
